main/uart_nic.c

```diff
@@ -164,17 +193,56 @@ static void event_handler(void* arg, esp_event_base_t event_base, int32_t event_
         }
         ESP_LOGI(TAG,"connect to the AP fail");
     } else if (event_base == WIFI_EVENT && event_id == WIFI_EVENT_STA_CONNECTED) {
-        last_inbound_seen = now_seconds();
-        associated = true;
         send_link_status(1);
         s_retry_num = 0;
-    }
+        ESP_ERROR_CHECK(esp_wifi_set_inactive_time(ESP_IF_WIFI_STA, INACTIVE_BEACON_SECONDS));
+        probe_schedule();
+    } else if (event_base == WIFI_EVENT && event_id == WIFI_EVENT_SCAN_DONE) {
+        wifi_ap_record_t ap_info;
+        ESP_ERROR_CHECK(esp_wifi_sta_get_ap_info(&ap_info));
+
+        const wifi_event_sta_scan_done_t *scan_data = (const wifi_event_sta_scan_done_t *)event_data;
+        uint16_t ap_count = scan_data->number;
+
+        uint8_t found = 0;
+        if (!scan_data->status && ap_count) {
+            wifi_ap_record_t *aps = (wifi_ap_record_t *)malloc(sizeof(wifi_ap_record_t) * ap_count);
+            ESP_ERROR_CHECK(esp_wifi_scan_get_ap_records(&ap_count, aps));
+            // Try to match BSSID first and if that fails go on and try SSID match . The BSSD check
+            // should be sufficient, but there are APs that advertise mismatching BSSID in their
+            // beacons and/or probe rensonse. That's the real culprit of the beacon timeout
+            // disconnects and the primary motivation of this whole excercise.
+            for (int i = 0; i < ap_count; ++i) {
+                if (0 == memcmp(ap_info.bssid, aps[i].bssid, 6)) {
+                    found = 1;
+                    break;
+                }
+            }
+            if (!found) {
+                for (int i = 0; i < ap_count; ++i) {
+                    if (ap_info.ssid && ap_info.ssid[0] && aps[i].ssid && aps[i].ssid[0]) {
+                        if (0 == strncmp((char *)(ap_info.ssid), (char *)(aps[i].ssid), 32)) {
+                            found = 1;
+                            break;
+                        }
+                    }
+                }
+            }
+            free(aps);
+        }
+        if (!found){
+            if (probe_retry_count++ < probe_max_reties) {
+                probe_run();
```

Comment:
Time is irrelevant. The AP's gone. If we can't see the beacon which is being broadcast at the lowest bitrate on the lowest channel, the data connection is 100% dead.

Reply:
I'm thinking about a different thing. The scan retry might take short or long time. Therefore, the detection of being down might take 2 seconds or 25… and we don't really know which. That's a bit unpredictable to the user, so I wonder if a better / more predictable way is to keep trying the scans for some time instead of 3 times (the same argument, that it's surely dead, could also be applied to the retries).